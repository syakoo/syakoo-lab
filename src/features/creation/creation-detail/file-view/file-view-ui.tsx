"use client";

import { useState } from "react";
import type { FC, PropsWithChildren } from "react";

import { Icon } from "@/shared/design-system/icons";
import { Text } from "@/shared/design-system/ui";

import { styles } from "./file-view.css";

type Props = PropsWithChildren<{
  fileName: string;
}>;

export const FileViewUI: FC<Props> = ({ children, fileName }) => {
  const [isOpen, setIsOpen] = useState(false);

  const handleToggle = () => {
    setIsOpen((prev) => !prev);
  };

  return (
    <div className={styles.root}>
      <button className={styles.summary} onClick={handleToggle} type="button">
        <div className={styles.icon({ open: isOpen })}>
          <Icon height={16} name="chevron-right" width={16} />
        </div>
        <Text size="75">{fileName}</Text>
      </button>
      {isOpen ? <div className={styles.content}>{children}</div> : null}
    </div>
  );
};
