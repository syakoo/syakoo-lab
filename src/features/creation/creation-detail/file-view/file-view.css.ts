import { style } from "@vanilla-extract/css";
import { recipe } from "@vanilla-extract/recipes";

import { theme } from "@/shared/design-system/theme.css";

const icon = recipe({
  base: {
    transition: "transform 0.2s ease-in-out",
    display: "flex", // アイコンを中央寄せにするため
    alignItems: "center", // アイコンを中央寄せにするため
  },
  variants: {
    open: {
      true: {
        transform: "rotate(90deg)",
      },
    },
  },
});

export const styles = {
  root: style({}),

  summary: style({
    display: "flex",
    alignItems: "center",
    gap: theme.space[25],
    width: "100%",
    padding: theme.space[50],
    cursor: "pointer",
    color: theme.color.text.secondary,

    ":hover": {
      color: theme.color.text.primary,
    },
  }),

  content: style({
    border: `1px solid ${theme.color.palette.gray[200]}`,
    borderRadius: "4px",
    padding: theme.space[100],
    whiteSpace: "pre-wrap",
    wordBreak: "break-all",
  }),

  icon,
};
