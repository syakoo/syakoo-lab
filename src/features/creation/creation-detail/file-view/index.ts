export { FileView } from "./file-view";
