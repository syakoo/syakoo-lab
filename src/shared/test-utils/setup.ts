import { exampleMDXComponent } from "@/features/mdx/mocks/fixture-mdxcomponent";

import { loadEnv } from "./nextjs/env";

// NOTE: `@mdx-js/mdx` の esm がうまく解決できなかったためまとめてモック化
// これが原因でおかしくなっていたらごめんなさい
jest.mock("@mdx-js/mdx", () => {
  return {
    runSync: jest
      .fn()
      .mockImplementation(() => ({ default: exampleMDXComponent })),
  };
});

// NOTE: `mermaid` の esm がうまく解決できなかったためまとめてモック化
// これが原因でおかしくなっていたらごめんなさい
jest.mock("mermaid", () => ({
  initialize: () => {
    return;
  },
  run: () => {
    return;
  },
}));

loadEnv();

jest.spyOn(Math, "random").mockImplementation(() => 0.5866245602626035);
