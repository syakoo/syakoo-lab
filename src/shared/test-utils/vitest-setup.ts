import { vi } from "vitest";

import { exampleMDXComponent } from "@/features/mdx/mocks/fixture-mdxcomponent";

import { loadEnv } from "./nextjs/env";

// NOTE: Next.jsのImageコンポーネントをテスト用にモック
vi.mock("next/image", () => ({
  default: "img",
}));

// NOTE: `@mdx-js/mdx` の esm がうまく解決できなかったためまとめてモック化
// これが原因でおかしくなっていたらごめんなさい
vi.mock("@mdx-js/mdx", () => {
  return {
    runSync: vi
      .fn()
      .mockImplementation(() => ({ default: exampleMDXComponent })),
  };
});

// NOTE: `mermaid` の esm がうまく解決できなかったためまとめてモック化
// これが原因でおかしくなっていたらごめんなさい
vi.mock("mermaid", () => ({
  default: {
    initialize: () => {
      return;
    },
    run: () => {
      return;
    },
  },
  initialize: () => {
    return;
  },
  run: () => {
    return;
  },
}));

loadEnv();

vi.spyOn(Math, "random").mockImplementation(() => 0.5866245602626035);
