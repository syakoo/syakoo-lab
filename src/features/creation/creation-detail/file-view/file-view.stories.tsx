import type { Meta, StoryObj } from "@storybook/nextjs";

import { FileViewUI } from "./file-view-ui";

const meta = {
  component: FileViewUI,
  parameters: {},
} satisfies Meta<typeof FileViewUI>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    fileName: "example.txt",
    children: "This is the content of the file.",
  },
};

export const LongFileName: Story = {
  args: {
    fileName: "this_is_a_very_long_file_name_that_should_be_truncated.txt",
    children: "This is the content of the file.",
  },
};
