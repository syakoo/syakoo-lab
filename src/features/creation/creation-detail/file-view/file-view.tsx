import { promises as fs } from "fs";
import path from "path";

import type { FC } from "react";

import { FileViewUI } from "./file-view-ui";

type Props = {
  src: string;
  fileName?: string;
};

export const FileView: FC<Props> = async ({ src, fileName }) => {
  const filePath = path.join(process.cwd(), "public", src);
  const content = await fs.readFile(filePath, "utf-8");
  const displayFileName = fileName ?? path.basename(src);

  return <FileViewUI fileName={displayFileName}>{content}</FileViewUI>;
};
