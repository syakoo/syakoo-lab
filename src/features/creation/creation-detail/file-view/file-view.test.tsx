import { testStories } from "@/shared/test-utils/test-stories";

import * as stories from "./file-view.stories";

describe("FileView", () => {
  testStories(stories);
});
