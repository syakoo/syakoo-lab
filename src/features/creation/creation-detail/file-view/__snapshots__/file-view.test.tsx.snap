// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FileView > スナップショット Default 1`] = `
<body>
  <div>
    <div
      class="file-view_styles_root__153z7ur2"
    >
      <button
        class="file-view_styles_summary__153z7ur3"
        type="button"
      >
        <div
          class="file-view_icon__153z7ur0"
        >
          <svg
            fill="none"
            height="16"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m8.25 4.5 7.5 7.5-7.5 7.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
        <p
          class="text_textStyle__1djq151e text_textStyle_color_currentColor__1djq151i text_textStyle_variants_color_currentColor__1djq1513 text_textStyle_size_75__1djq151k text_textStyle_variants_size_75__1djq1515 text_textStyle_weight_normal__1djq151u"
        >
          example.txt
        </p>
      </button>
    </div>
  </div>
</body>
`;

exports[`FileView > スナップショット LongFileName 1`] = `
<body>
  <div>
    <div
      class="file-view_styles_root__153z7ur2"
    >
      <button
        class="file-view_styles_summary__153z7ur3"
        type="button"
      >
        <div
          class="file-view_icon__153z7ur0"
        >
          <svg
            fill="none"
            height="16"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m8.25 4.5 7.5 7.5-7.5 7.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
        <p
          class="text_textStyle__1djq151e text_textStyle_color_currentColor__1djq151i text_textStyle_variants_color_currentColor__1djq1513 text_textStyle_size_75__1djq151k text_textStyle_variants_size_75__1djq1515 text_textStyle_weight_normal__1djq151u"
        >
          this_is_a_very_long_file_name_that_should_be_truncated.txt
        </p>
      </button>
    </div>
  </div>
</body>
`;
